internal/database/autocontact_test.go

```diff
@@ -0,0 +1,152 @@
+package database
+
+import (
+	"log/slog"
+	"os"
+	"testing"
+	"time"
+
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+)
+
+func TestSSHRestart(t *testing.T) {
+	// Setup
+	logger := slog.Default()
+	sshConfig := SSHConfig{
+		User:    "testuser",
+		Keypath: "/path/to/your/test/private/key",
+	}
+
+	err := sshRestart("invalid.remote.address", logger, sshConfig)
+	if err == nil {
+		t.Error("Expected sshRestart to fail due to invalid setup, but it did not")
+	} else {
+		t.Log("SSHRestart attempted as expected:", err)
+	}
+
+}
+
+func TestSSHRestart_UnreadableKey(t *testing.T) {
+	logger := slog.Default()
+	sshConfig := SSHConfig{
+		User:    "testuser",
+		Keypath: "/path/to/nonexistent/key", // Intentionally wrong to trigger the read error
+	}
+
+	err := sshRestart("localhost", logger, sshConfig)
+	if err == nil {
+		t.Error("Expected an error due to unreadable key, but got none")
+	} else {
+		t.Log("Received expected error:", err)
+	}
+}
+
+func TestSSHRestart_UnparsableKey(t *testing.T) {
+	logger := slog.Default()
+	// Create a temp file with invalid SSH key content
+	tmpFile, err := os.CreateTemp("", "invalid-ssh-key-*")
+	if err != nil {
+		t.Fatalf("Failed to create temp file: %v", err)
+	}
+	defer os.Remove(tmpFile.Name()) // clean up
+
+	_, err = tmpFile.WriteString("not-a-valid-key")
+	if err != nil {
+		t.Fatalf("Failed to write to temp file: %v", err)
+	}
+	tmpFile.Close()
+
+	sshConfig := SSHConfig{
+		User:    "testuser",
+		Keypath: tmpFile.Name(), // Use the temp file with invalid content
+	}
+
+	err = sshRestart("localhost", logger, sshConfig)
+	if err == nil {
+		t.Error("Expected an error due to unparsable key, but got none")
+	} else {
+		t.Log("Received expected error:", err)
+	}
+}
+
+func TestPing(t *testing.T) {
+	logger := slog.Default()
+
+	success := ping("192.0.2.1", logger)
+
+	if success {
+		t.Error("Expected ping to fail for a non-routable IP address, but it reported success")
+	} else {
+		t.Log("Ping processed as expected")
+	}
+}
+
```

Comment:
This needs a test for the happy path, as it's fragile, and doesn't work. This should either go to loopback, or google.com.


```go
func TestPingHappy(t *testing.T) {
	logger := slog.New(slog.NewTextHandler(os.Stderr, &slog.HandlerOptions{
		Level: slog.LevelDebug,
	}))

	success := ping("google.com", logger)

	if !success {
		t.Error("Failed to ping google")
	}
}
```

```
alona@Ashtabula:~/dev/gpuctl$ go test -v -run TestPingHappy ./internal/database/
=== RUN   TestPingHappy
time=2024-02-15T22:23:13.550Z level=DEBUG msg="Error listening for ICMP packets:" !BADKEY="listen ip4:icmp 0.0.0.0: socket: operation not permitted"
    autocontact_test.go:95: Failed to ping google
--- FAIL: TestPingHappy (0.02s)
FAIL
FAIL    github.com/gpuctl/gpuctl/internal/database      0.027s
FAIL
```

Reply:
FYI I think the only reason this doesn't work because you don't have perms to make raw sockets, and so you should test with sudo. The new implementation, with shell-based pinging, passes this test.